### PR TITLE
TE-11.3  implemented changes as per PR 1158

### PR DESCRIPTION
**Pull request by bkreddy143**
Implement the updated test plan per #1158 

1. Added port 4 to topology and modified vrf names
2. Added port 1 to VRF-1
2. Added Decap and Encap support
3. Added backup support primary to decap+encap and decap+encap to decap
4. Added packet validation based on path
5. Added lookup vrf support in AddNH funtion and also optimized the code This code is a Contribution to the OpenConfig 
6. Also merged README.md file from original PR#1158
Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind.

### CODE REVIEW

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go
Comment (cprabha) [inline feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go:195]: Default network instance should have config type as "DEFAULT_INSTANCE".

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go:137]: Read me will get updated as part of PR-1158 once that is merged.

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go
Comment (xw-g) [inline feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go:32]: Shall we make the variable names as `vrfA`, `vrfB` and `vrfC` for the easy of reading?

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go
Comment (xw-g) [inline feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go:324]: Same comment about log message as above.

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go
Comment (xw-g) [inline feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go:342]: Typo in the log message? NHG100 does not do decap.

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go
Comment (LimeHat) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go:195]: @cprabha this is not related to the lines your highlighted. 
Changes that were introduced as a part of #1335 will be added during branch merge with the upstream main; this pr is branched off #1158 and will be merged into it.

**Review by xw-g — COMMENTED** [review]
If it makes things convenient for you, I'm ok with copy&paste PR#1158 README here and close PR#1158, so that you can directly pull into main.  (But you need to mention PR#1158 in your PR description). 

Also, need to update OTG test. Thanks.

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go
Comment (cprabha) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go:137]: @xw-g , check PR # 1158 readme looks good.

**Review by greg-dennis — CHANGES_REQUESTED** [review]

**Comment by cprabha** [comment]
> @bkreddy143 thanks for the contribution. This is good. Overall looks good to me, just some minor comment.
> 
> @cprabha can you help review too? Any other concerns? (non-blocking or vendor specific changes can come after later with incremental PRs)

@xw-g , sure, will check it. Thank you

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go
Comment (cprabha) [inline feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go:364]: Could you please move the programming of NH104, NHG104 and outerDstIP2 (L#364 - L#369) before programming NH102 (L#356) ? This is to ensure "Dest: outerDstIP2" is programmed before using it.

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go]: Updated IPs as per guidelines.

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go]: Update with small case variable names

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go]: Updated

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go]: updated

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go]: corrected

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go]: updated

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go]: updated

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go]: corrected

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go]: updated

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go]: Corrected

File: internal/gribi/gribi.go
Comment (bkreddy143) [reply internal/gribi/gribi.go]: @greg-dennis  are you taking about PR-1181 ?  In this PR i have optimized the code and also added validation for each case where as existing code validate modify response only for default case .  Also update variable to smaller case.

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go:137]: RAEDME.md  changes from PR-1158 are merged to this PR and will be closing 1158

**Comment by xw-g** [comment]
@cprabha approve?

File: internal/gribi/gribi.go
Comment (LimeHat) [inline internal/gribi/gribi.go:207]: Let's use `VRFOnly` to match the updated te-11.1 and te-11.2 tests (seems all three are about to be merged)

**Review by cprabha — APPROVED** [review]
Looks good.

**Comment by xw-g** [comment]
@mojiiba ok to merge? If any deviation needed or vendor specific, let's use other PRs to iterate. Thanks.

**Comment by vbhan-cisco** [comment]
You can merge this PR. Moji is away today

**Comment by xw-g** [comment]
@vosipchu also just gave LGTM. 

Given that this is a change of test plan and we have many reviewers already agreed with the logic and code changes, I'll go ahead merge this PR.

**Review by xw-g — DISMISSED** [review]

**Comment by xw-g** [comment]
@bkreddy143 can you help address the merge conflict? Thanks.

**Review by greg-dennis — APPROVED** [review]

File: internal/gribi/gribi.go
Comment (bkreddy143) [reply internal/gribi/gribi.go]: Updated

**Comment by bkreddy143** [comment]
> @bkreddy143 can you help address the merge conflict? Thanks.
> 
> I think the merge conflict probably is because this gribi.go file: https://github.com/openconfig/featureprofiles/pull/1181/files#diff-e495bef600e4cd54f409c7546966272dd971aed71b6d649251d3eb16cf2c4b78.
> 
> Let me know if this is not the case, and the conflict actually changed the test case and requires another review. Thanks.


@xw-g  , I resolved conflicts .

**Review by xw-g — APPROVED** [review]

**Pull request merged**
